### ISSUE TITLE
DefaultSecurityLogic can only work with DirectClients? 

### ISSUE THREAD

**Issue by mmoayyed** [pac4j/pac4j#751]
There exists this line:

```
if (currentClient instanceof DirectClient) {
```

If the currentClient is a CasClient, a profile can never be established. Requests always fail and are redirected to the identity provide again and again. This seems like a bug?

**Comment by leleuj** [comment]
No, this is really intended.

Either you want to protect a web service and you use a direct client (the authentication and profile retrieval will happen during the request) or you want to protect a UI and you use an indirect client (the requested URL is saved, the user is redirected to the identity provider for login, the login process is finished on the callback URL, the originally requested URL is restored).

For CAS, if you use the `CasClient`, you protect a UI (and add a `CasProxyReceptor` to handle proxy tickets creation). Or you can use the `DirectCasproxyClient` to protect a web service and validate proxy tickets. Or you can use the `CasRestBasicAuthClient` or the `CasRestFormClient` to protect a web service and validate the credentials via the CAS REST API.

Isn't the documentation clear enough? -> http://www.pac4j.org/1.9.x/docs/clients/cas.html

**Comment by mmoayyed** [comment]
No none of those formulas work, because:

- Using CasClient means I'll be using an IndirectClient
- Using a SecurityInterceptor with the DefaultSecurityLogic requires I use a DirectClient

Is that not broken?

**Comment by mmoayyed** [comment]
Also note that from my perspective, the check is rather arbitrary? Why is the default security logic only functional with direct clients?

**Comment by leleuj** [comment]
Not exactly, `IndirectClient`s are also handled by the security filter, but not in the same way: if you define an indirect client, the authentication will not be tried directly, but the user will be redirected to the identity provider for login and the login process will happen on the callback URL for the indirect client.

To sum up:
1) the security filter:
- protects an URL
- performs the login process for a direct client or redirect to the identity provider for an indirect client
2) the callback filter:
- performs the login process for an indirect client.

**Comment by mmoayyed** [comment]
Who defined the callback filter? and why does it have to be different than the security filter?

**Comment by mmoayyed** [comment]
And why does the security logic by default makes an assumption about the client type? where does that rationale come from?

**Comment by leleuj** [comment]
The user defines the callback URL if he needs it (= if he has indirect clients in his security configuration).

Conceptually, they are be different: one protects the URL, one finishes the login process.
Practically, it's fairly complicated to mix both because we need to know if it's a callback or a first call which requires a redirection.
Currently, the security model is clear (at least, that is my feeling).

I'm not sure to understand your last comment: the security behavior depends on the type of the client.

Let's take the issue the other way: what can't you do with the current security model?

**Comment by mmoayyed** [comment]
Sure thing. Here is my concrete use case:

The management webapp in CAS defines a CasClient instance that is put wired up by the SecurityInterceptor:
https://github.com/apereo/cas/blob/master/webapp-mgmt/cas-management-webapp-support/src/main/java/org/apereo/cas/mgmt/config/CasManagementWebAppConfiguration.java#L129

Fairly simple. 

When users return from the IdP (CAS), the interceptor kicks to locate the client and validate the request to build the profile.
The DefaultSecurityLogic can never do this because it has that instanceof check that prohibits indirect clients to do their job. 
Which is why you see the current CAS config in the state in is (defined with a DirectClient). 

Using CasClient results in redirect loops because the SecurityLogic has this: `if (currentClient instanceof DirectClient) {`. I am not sure why it ever needs to make that assumption, apart from the fact that this seems broken somehow? You can try to duplicate this issue with the management webapp and swap out the config for a CasClient and see the behavior. It's entirely possible something trivial is broken on the CAS end but this somehow has become needless more confusing that it should be.

**Comment by leleuj** [comment]
The `CasClient` is different from the `DirectCasClient`.

The `CasClient` is an indirect client whose callback endpoint is the regular static callback URL where the service ticket validation will happen.
In that case, the ST validation will happen in the `CallbackController`, not in the `SecurityInterceptor`.

Simplified flow (stateful):

-> /protectedurl
-> /cas/login?service=/callback
-> /callback?service=xxx&ticket=yyy
-> /protectedurl

---

The `DirectCasClient` is a specific direct client for the CAS server and the OAuth server support, to be stateless.
It's behaviour is easy: on a protected URL, either you have a ST and the validation is performed or you don't have a ST and the user is redirected to the CAS server for login (and the service is the requested URL).
See: https://github.com/pac4j/pac4j/blob/master/pac4j-cas/src/main/java/org/pac4j/cas/client/direct/DirectCasClient.java#L60

Simplified flow (stateless):

-> /protectedurl
-> /cas/login?service=/protectedurl
-> /protectedurl?service=xxx&ticket=yyy

For the management webapp, you should definitely use the `CasClient`, a `SecurityInterceptor` and a `CallbackController`. I think it's what I did first when I migrated to `spring-webmvc-pac4j`.

**Comment by mmoayyed** [comment]
Thanks for the thoughtful walkthrough. Few feedback items:

- this whole debacle of direct vs indirect is terribly confusing to me. Long term it's a lot more reasonable to have a single client abstraction. It's less work for everyone. 
- I would much prefer to use the "stateless" route. It's much simpler to setup and it's more natural with cas. I don't need nor want to setup a separate callback endpoint. You can already see there one less step to do.
- while your description is adequate, i dont see what the security logic needs to make assumptions about the client. That's needless coupling. I understand the interceptor is designed to work with direct client but that option I am not sure needs to be coded as a condition.

**Comment by mmoayyed** [comment]
In simpler terms: the narrative should be that I have a security interceptor who handles security for my app. It redirects me to my idp and the handles the response when I return and let's me in. It's a gatekeeper and should be. All other complications of this story are pure technical noise and should be removed where and when possible.

**Comment by mmoayyed** [comment]
Also what exactly is stateless about the 2nd approach? Because we are not keeping and remembering the profile in the request or the session? I don't think that's the case.

**Comment by leleuj** [comment]
I understand your feelings, going through all of this seems unnecessarily complicated, but I don't think so. When you gather all the constraints and use cases we want to handle, I don't think we can really do better.

Spring Security and Shiro for example, are designed on concepts then authentication methods are implemented thanks to these concepts.
pac4j is designed from two use cases: web services or UI authentication -> direct or indirect clients. They are different by definition.

Now the constraints:

1) Why you can't be stateless (stateless means we don't remember (save / restore in session) the requested URL)

For OpenID Connect and OAuth, all identity providers allow you to define only one or several callback URLs, which means you can't use the current requested URL as a callback URL. So you need some unique URL for callback and be able to remember the requested URL -> stateful

2) Why the SecurityInterceptor can't play both roles for indirect clients

On the `SecurityInterceptor`, if we are not authenticated, we need to start a login process (for Facebook for example) and we can use the `client_name` parameter to choose an authentication method.
On the `CallbackController`, given the `client_name`, we finish the appropriate login process. The user is not authenticated.
If the callback logic is put into the `SecurityInterceptor`, when the user is not authenticated, how can I know that I need to trigger the login process or finish it?

In fact, we could: just put a counter in session: nothing in session -> login process (+1 in session), 1 in session -> finish the process (0 in session), more than 1 -> unexpected loop / unauthorized

So at best, we could melt the `SecurityInterceptor` and the `CallbackController` with some internal complexity, but direct clients and indirect clients remain different.
That said, we can't escape the static callback URL, so, following your reasoning, we would need to protect it by the `SecurityInterceptor` instead of handling it with the anonymous `CallbackController`: I'm not convinced it's an improvement given the philosophical breaking change.

But I'm open to your feedbacks.

**Comment by mmoayyed** [comment]
Thanks so much for the excellent explanation. Let me have some time to digest this all; I think I can provide a few snippets on what I have in mind to perhaps simplify this. Stay tuned.

**Comment by mmoayyed** [comment]
Going to leave this open under discussion. Hope that's OK (or we can take this to dev if you prefer though sharing code snippets is easier here)

**Comment by leleuj** [comment]
No problem. If you have some improvements, let's share it first on this issue.

**Comment by mmoayyed** [comment]
> So at best, we could melt the SecurityInterceptor and the CallbackController with some internal complexity, but direct clients and indirect clients remain different.

I think this is really my final conclusion. I had a chance to go through this. I think I do have a solution where `SecurityInterceptor` is melted into the `CallbackController`. The entire flow is handled with one component altogether, and there is no need to set up anything separately. Specially in the CAS world, the current behavior is rather unorthodox. I'll see if I can share a few snippets to discuss the change here before formally turning it into a PR.

**Comment by leleuj** [comment]
I'm looking forward to see your proposals...

**Comment by mmoayyed** [comment]
How about?

```java
public class TestSecurityInterceptor extends SecurityInterceptor {

    public TestSecurityInterceptor() {
        super(config(), "DirectCasClient", "securityHeaders,csrfToken");
        final DefaultSecurityLogic logic = new DefaultSecurityLogic() {
            @Override
            protected HttpAction forbidden(final WebContext context, final List currentClients, final List list, final String authorizers) {
                return HttpAction.redirect("Authorization failed", context, "authorizationFailure");
            }

            @Override
            protected boolean loadProfilesFromSession(final WebContext context, final List currentClients) {
                return true;
            }
        };

        logic.setSaveProfileInSession(true);
        setSecurityLogic(logic);
    }

    @Override
    public void postHandle(final HttpServletRequest request, final HttpServletResponse response,
                           final Object handler, final ModelAndView modelAndView) throws Exception {
        final RedirectView v = new RedirectView(request.getRequestURL().toString());
        v.setExposeModelAttributes(false);
        v.setExposePathVariables(false);
        modelAndView.setView(v);
    }
}
```

**Comment by mmoayyed** [comment]
I should point out that the interceptor is mapped to `/**`.

**Comment by leleuj** [comment]
OK. That's a good example: you use the `DirectCasClient` which is stateless (the user profile is not saved into session), so you are forced to override the loading and saving with the web session.

But if you have a web session, why not simply use the `CasClient`?

**Comment by mmoayyed** [comment]
Mostly because this model with the DirectCasClient is much simpler to setup. The alternative is to setup a callback filter and massage the interceptor to exclude that bit, etc. The setup in comparison requires more effort. Also note that DirectCasClient is more in alignment with how most CAS clients work: 

Narrative in plain english is:
> You start with A -> You get redirected to CAS with A -> CAS redirects back to A -> A validates. -> Happy face.

It's rather unorthodox where you are asked to switch out A with a callback url that eventually redirects back to A. I understand why that has to happen. I just hope to avoid it if I can because it's purely a technical complication and not quite "added value". I appreciate the nuances of what direct vs indirect attempts to accomplish. By all means. Sure. What I am wondering is, if there is anything we can do here based on my simple snippet, etc to simplify the setup and steps it takes to accomplish the simplest use case.

**Comment by mmoayyed** [comment]
Of course if the consensus is to not allow for this sort of "flexibility" and expect more understanding from the developer/adopter, that's not a problem either as long as we understand that's our position going forward.

**Comment by leleuj** [comment]
I'm starting to get your point of view. No specific callback -> no `CallbackController`, just the `SecurityInterceptor`. Though, this is not such a cumbersome process to declare both components. 

For CAS, you're right, we can use regular expressions to define the service URLs so it's easy to avoid the callback URL.
That said, it's more an exception than a general rule: for SAML, OAuth and OIDC, you need to define a specific callback URL.
And I don't know all CAS clients, but some of them use a specific callback URL like Spring Security.

I'm used to the "callback endpoint for indirect clients", I think it makes it clear there will be a callback from the identity provider, but I may be biased on this after ingesting my own model for years.

**Comment by mmoayyed** [comment]
Furthermore, this line is also very strange in the same context:

```java
currentClients.get(0) instanceof IndirectClient;
```

Why do we have to assume [and check] that the first client in the list should be an indirect client? Might be better if the list of clients are scanned to find the first immediate indirect client and use that. If more than 1 is found, error out with an explicit error message.

**Comment by leleuj** [comment]
It's a convention: on the security filter, direct clients are tried successfully. If the user is not authenticated, we check the first client and if it's an indirect one, we redirect to it for login.

We may have several indirect clients defined to be able to choose one of them if necessary.

Example:
`clients` = `FacebookClient,TwitterClient,DirectBasicAuthClient`

When calling `/protected` and not authenticated and no basic auth, the user is redirected to Facebook
When calling `/protected?client_name=TwitterClient` and not authenticated and no basic auth, the user is redirected to Twitter.

**Comment by mmoayyed** [comment]
This was a great discussion here; thank you for your time! I don't feel like at this point there is anything we can add/fix per se, so I am going to close. If I bump into anything that I feel deserves a change, I shall reopen and continue the works.

**Comment by leleuj** [comment]
No problem. It's good to challenge the decisions I made and to try to improve things.
